smart_open/tests/test_gcs.py

```diff
@@ -0,0 +1,660 @@
+    def __init__(self, credentials=None):
+        if credentials is None:
+            credentials = FakeCredentials(self)
+        self._credentials = credentials  # type: FakeCredentials
+        self._uploads = {}
+        self.__buckets = []
+
+    def bucket(self, bucket_id):
+        bucket = next((x for x in self.__buckets if x.name == bucket_id), None)
+        if bucket is None:
+            raise google.cloud.exceptions.NotFound('Bucket {} not found'.format(bucket_id))
+        return bucket
+
+    def create_bucket(self, bucket_id):
+        bucket = FakeBucket(self, bucket_id)
+        self.__buckets.append(bucket)
+        return bucket
+
+    def get_bucket(self, bucket_id):
+        return self.bucket(bucket_id)
+
+    def _delete_bucket(self, bucket):
+        self.__buckets.remove(bucket)
+
+
+class FakeBlobUpload(object):
+    def __init__(self, url, blob):
+        self.url = url
+        self.blob = blob  # type: FakeBlob
+
+
+class FakeResponse(object):
+    def __init__(self, status_code=200):
+        self.status_code = status_code
+
+
+class FakeAuthorizedSession(object):
+    def __init__(self, credentials):
+        self._credentials = credentials  # type: FakeCredentials
+
+    def delete(self, upload_url):
+        blob = self._credentials.client._uploads[upload_url]
+        blob.delete()
+        self._credentials.client._uploads.pop(upload_url)
+
+    def put(self, url, data=None, headers=None):
+        if data is not None:
+            self._credentials.client._uploads[url].write(data.read())
+        return FakeResponse()
+
+    def _blob_with_url(self,
+                       url,
+                       client  # type: FakeClient
+                       ):
+        return client._uploads.get(url)
+
+
+if DISABLE_MOCKS:
+    storage_client = google.cloud.storage.Client()
+else:
+    storage_client = FakeClient()
+
+
+def get_bucket():
+    return storage_client.bucket(BUCKET_NAME)
+
+
+def get_blob():
+    bucket = get_bucket()
+    return bucket.blob(BLOB_NAME)
+
+
+def cleanup_bucket():
+    bucket = get_bucket()
+
+    blobs = bucket.list_blobs()
+    for blob in blobs:
+        blob.delete()
+
+
+def put_to_bucket(contents, num_attempts=12, sleep_time=5):
+    logger.debug('%r', locals())
+
+    #
+    # In real life, it can take a few seconds for the bucket to become ready.
+    # If we try to write to the key while the bucket while it isn't ready, we
+    # will get a StorageError: NotFound.
+    #
+    for attempt in range(num_attempts):
+        try:
+            blob = get_blob()
+            blob.upload_from_string(contents)
+            return
+        except google.cloud.exceptions.NotFound as err:
+            logger.error('caught %r, retrying', err)
+            time.sleep(sleep_time)
+
+    assert False, 'failed to create bucket %s after %d attempts' % (BUCKET_NAME, num_attempts)
+
+
+def mock_gcs(func):
+    '''Mocks the function and provides additional required arguments.'''
```

Comment:
```suggestion
    '''Mock the function and provide additional required arguments.'''
```

PEP257: Docstrings in imperative mode.